contracts/javascore/modules/ics20app/src/main/java/ibc/ics20app/ICS20Transfer.java

```diff
@@ -0,0 +1,214 @@
+package ibc.ics20app;
+
+import ibc.icon.interfaces.IIBCModule;
+import ibc.icon.score.util.StringUtil;
+import ibc.ics23.commitment.Ops;
+
+import icon.proto.core.channel.Channel;
+import icon.proto.core.channel.Packet;
+import score.Address;
+import score.Context;
+import score.DictDB;
+import score.annotation.External;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+
+import static ibc.ics20app.ICS20TransferBank.bank;
+
+public abstract class ICS20Transfer implements IIBCModule {
+    public static final String ICS20_VERSION = "ics20-1";
+    public static final Address ZERO_ADDRESS = Address.fromString("hx0000000000000000000000000000000000000000");
+    public static final DictDB<String, Address> channelEscrowAddresses = Context.newDictDB("channelEscrowAddresses", Address.class);
+    protected final DictDB<String, String> destinationPort = Context.newDictDB("destinationPort", String.class);
+    protected final DictDB<String, String> destinationChannel = Context.newDictDB("destinationChannel", String.class);
+
+    @External(readonly = true)
+    public Address getIBCAddress() {
+        return ICS20TransferBank.ibcHandler.getOrDefault(ZERO_ADDRESS);
+    }
+
+    public void onlyIBC() {
+        Context.require(Context.getCaller().equals(getIBCAddress()), "ICS20App: Caller is not IBC Contract");
+    }
+
+    @External(readonly = true)
+    public String getDestinationPort(String channelId) {
+        return destinationPort.get(channelId);
+    }
+
+    @External(readonly = true)
+    public String getDestinationChannel(String channelId) {
+        return destinationChannel.get(channelId);
+    }
+
+    @External
+    public byte[] onRecvPacket(byte[] packet, Address relayer) {
+        onlyIBC();
+        Packet packetDb = Packet.decode(packet);
+        ICS20Lib.PacketData data = ICS20Lib.unmarshalJSON(packetDb.getData());
+        boolean success = _decodeReceiver(data.receiver);
+        if (!success) {
+            return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+        Address receiver = Address.fromString(data.receiver);
+
+        byte[] denomPrefix = getDenomPrefix(packetDb.getSourcePort(), packetDb.getSourceChannel());
+        byte[] denom = data.denom.getBytes();
+
+        if (denom.length >= denomPrefix.length && Ops.hasPrefix(denom, denomPrefix)) {
+            byte[] unprefixedDenom = Arrays.copyOfRange(denom, denomPrefix.length, denom.length);
+            String unprefixedDenomString = new String(unprefixedDenom);
+            if (unprefixedDenomString.equals("icx")){
+                success = _transferICX(receiver, data.amount);
```

Comment:
Since we are using the same escrow address for all tokens does this mean if one connection is malicious it can steal all ICX for all channels? 
I think this is what the generated escrow addresses are there for. We might need to account for this and perhaps track it serperatly